gradle-language-server/src/main/java/com/microsoft/gradle/GradleServices.java

```diff
@@ -102,6 +109,8 @@ private void compile(URI uri, GradleCompilationUnit unit) {
     Set<PublishDiagnosticsParams> diagnostics = new HashSet<>();
     try {
       unit.compile(Phases.CANONICALIZATION);
+      // If fatal error occurs and no AST generated, the visitCompilationUnit() will be ignored.
+      this.visitor.visitCompilationUnit(uri, unit);
```

Comment:
this ASTVisitor is to calculate Semantic Tokens. but `compile(uri, unit)` is shared among other tasks. So, will there be unnecessary calculation here e.g. when processing 'textdocument/completion` requests?

Reply:
We have stored the results in the ASTVisitor and get them via `this.visitor.getXXX()`. So if we support completion, we'll add some operations in the ASTVisitor and storage the completion related results. The approach is that we don't have to visit the AST more than once to the all the results. (for semantic highlighting, document symbol and completion)

If we separate the visitor for each usage, we might visit the AST more than once.